cgmanager.c

```diff
@@ -1114,6 +1131,253 @@ int list_children_main(void *parent, char *controller, const char *cgroup,
 	return get_directory_children(parent, path, output);
 }
 
+static int autoremove_entry_destroy(struct autoremove_entry *entry)
+{
+	nih_assert(entry != NULL);
+
+	nih_assert(entry->gpath != NULL);
+	nih_discard(entry->gpath);
+
+	nih_assert(entry->evpath != NULL);
+	nih_discard(entry->evpath);
+
+	nih_assert(entry->dirname != NULL);
+	nih_discard(entry->dirname);
+
+	if (entry->io != NULL)
+		nih_discard(entry->io);
+
+	nih_list_destroy(&entry->entry);
+
+	return 0;
+}
+
+/* if this function returns true then discontinue watching the events file */
+static bool autoremove_events_modified(struct autoremove_entry *entry)
+{
+	FILE *f;
+	char line[1024];
+	int pop_val = -1;
+	nih_local char *leafpath = NULL;
+
+	nih_assert(entry != NULL);
+	nih_assert(entry->evpath != NULL);
+
+	f = fopen(entry->evpath, "r");
+	if (!f) {
+		nih_error("%s: Cannot open %s in watcher: %s",
+			  __func__, entry->evpath, strerror(errno));
+		return false;
+	}
+
+	while (fgets(line, 1024, f) != NULL) {
+		int val;
+
+		if (sscanf(line, " populated %d", &val) != 1)
+			continue;
+
+		pop_val = !!val;
+		break;
+	}
+
+	fclose(f);
+
+	if (pop_val == -1) {
+		nih_error("%s: Cannot find or parse populated value in %s",
+			  __func__, entry->evpath);
+		return false;
+	} else if (pop_val == 1)
+		return false;
+
+	nih_assert(entry->gpath != NULL);
+	leafpath = NIH_MUST( nih_sprintf(NULL, "%s%s", entry->gpath, U_LEAF) );
```

Comment:
nih_sprintf allocates memory for the new string.  Since you are not assigning a parent, that means you need to manually nih_free(leafpath) right?

Reply:
leafpath is marked nih_local so it should get freed automatically when it goes out of scope (even
when its parent is NULL).

Reply:
Ah, right.  Thanks.